Shorten "helm-\(.*?\)-smart-do-search\(.*\)" to "hsearch-\1\2"

Bost opened the pull request:
With old names obsoleted for backward compatibility.

bcc32 commented:
What benefit does this have?  Sure, the names are shorter, but now they also obscure the fact that they are specific to helm.  Is `hsearch` a popular abbreviation for helm-related search functionality outside of Spacemacs?

Bost commented:
> What benefit does this have?

`SPC s` is bound to search/symbol functionality. When this keybinding is pressed, the function names are abbreviated, so it's difficult to identify the desired function. That's a poor user experience. See screenshots.

The original structure is `helm-<something>-smart-do-search-region-or-symbol`. The `search-<something>` is the most important piece of information. Next in importance is the scope, which is `region-or-symbol`. The fact that the search is done by 'helm-smart-do' is less critical, so I've replaced it with a single letter `h`.

My choice of names may not be optimal, so please feel free to propose better alternatives.
![spacemacs-hsearch-after](https://github.com/user-attachments/assets/c59a92a5-c689-42f2-ad2a-f2727b2d5de9)
![spacemacs-hsearch-before](https://github.com/user-attachments/assets/68bf81f3-6423-4e26-aa6e-9a7b9e1f4b40)

bcc32 commented:
How about fixing that in just the key bindings themselves, since this is primarily to affect which-key?  I feel like the new command names are a little better but still not that clear.  As an example:

```
      ;; current dir scope
      "sd"  '("search dir" . spacemacs/helm-dir-smart-do-search)
      "sD"  '("search dir w/ input" . spacemacs/helm-dir-smart-do-search-region-or-symbol)
```

Alternatively, if the other maintainers think renaming the commands is fine, we should perhaps rename to conform to the `helm/` layer prefix convention rather than `spacemacs/helm`

fnussbaum commented:
>How about fixing that in just the key bindings themselves, since this is primarily to affect which-key?

This is what I intended to suggest, too.

BostX committed:
[helm] Use explicit descriptions for keybinding menu items

Modify spacemacs||set-helm-key to accept (DESCRIPTION . FUNCTION) cons cells,
allowing human-readable menu text instead of auto-generated abbreviations from
long function names. This improves discoverability and consistency across helm
keybinding descriptions.

Bost commented:
Aaah, you're right. I reworked the patch. Please have a look.

bcc32 commented on layers/+completion/helm/funcs.el:
```diff
@@ -540,16 +540,17 @@ Removes the automatic guessing of the initial value based on thing at point. "
   "Define a key bindings for FUNC using KEYS.
 Ensure that helm is required before calling FUNC."
-  (let ((func-name (intern (format "lazy-helm/%s" (symbol-name func)))))
+  (let* ((actual-func (if (consp func) (cdadr func) func))
```
Since this macro expects unquoted function names in the normal case, I think it makes sense to also accept unquoted `("description" . func-name)` rather than with the quote.  Then this can just use `cdr` instead of `cdadr`---the latter is surprising.

Bost replied:
Done. BTW, we have a slight discrepancy in the quotation:
```
(spacemacs/set-leader-keys "<keys>"  '<func>)
(spacemacs/set-leader-keys "<keys>"  '("<desc>" . <func>))
```
vs.
```
(spacemacs||set-helm-key "<keys>"  <func>)
(spacemacs||set-helm-key "<keys>"  ("<desc>" . <func>))
```

It would be nice to have the same function signature for both `spacemacs||set-helm-key` and `spacemacs/set-leader-keys`.

bcc32 replied:
Yeah, agreed it would be more consistent.  I guess one reason it might be difficult is that `spacemacs||set-helm-key` needs its input to be a symbol and not just any old function (or perhaps a keymap, etc.).

bcc32 commented on layers/+completion/helm/packages.el:
```diff
@@ -250,34 +256,45 @@
-      "ss"  'spacemacs/helm-file-smart-do-search
-      "sS"  'spacemacs/helm-file-smart-do-search-region-or-symbol
+      "ss"  'spacemacs/hsearch-file
+      "sS"  'spacemacs/hsearch-file-region-or-symbol
```
I think this change needs to be reverted too?

Bost replied:
Yeah, sorry. That slipped through.

bcc32 reviewed (changes requested):
Mostly LGTM now, thanks for reworking the PR.  Made a couple small suggestions.

Bost commented:
Please have a look again.

bcc32 commented:
Looks good, thanks for your contribution!
